nifi-nar-bundles/nifi-hazelcast-bundle/nifi-hazelcast-services/src/main/java/org/apache/nifi/hazelcast/services/cachemanager/EmbeddedHazelcastCacheManager.java

```diff
@@ -0,0 +1,244 @@
+                        Long.valueOf(TimeUnit.SECONDS.toMillis(DEFAULT_CLIENT_BACKOFF_MAXIMUM_IN_SEC)).intValue(),
+                        DEFAULT_CLIENT_BACKOFF_MULTIPLIER);
+                getLogger().info("This host was not part of the expected Hazelcast instances. Hazelcast client has been started and joined to listed instances.");
+            }
+        } else if (clusteringStrategy.equals(CLUSTER_NONE.getValue())) {
+            result = Hazelcast.newHazelcastInstance(config);
+            getLogger().info(String.format(INSTANCE_CREATION_LOG, instanceName));
+        } else {
+            throw new ProcessException("Unknown Hazelcast Clustering Strategy!");
+        }
+
+        return result;
+    }
+
+    private List<String> getHazelcastMemberHosts(final PropertyContext context) {
+        return Arrays.asList(context.getProperty(HAZELCAST_INSTANCES).evaluateAttributeExpressions().getValue().split(ADDRESS_SEPARATOR))
+                .stream().map(i -> i.trim()).collect(Collectors.toList());
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTY_DESCRIPTORS;
+    }
+
+    @Override
+    protected Collection<ValidationResult> customValidate(final ValidationContext context) {
+        final Set<ValidationResult> results = new HashSet<>();
+
+        // This validation also prevents early activation: up to the point node is considered as clustered this validation step
+        // prevents automatic enabling when the flow stars (and the last known status of the controller service was enabled).
+        if (!getNodeTypeProvider().isClustered() && context.getProperty(HAZELCAST_CLUSTERING_STRATEGY).getValue().equals(CLUSTER_ALL_NODES.getValue())) {
+            results.add(new ValidationResult.Builder()
+                    .subject(HAZELCAST_CLUSTERING_STRATEGY.getName())
+                    .valid(false)
+                    .explanation("Cannot use \"All Nodes\" Clustering Strategy when NiFi is not part of a cluster!")
+                    .build());
+        }
+
+        if (!getNodeTypeProvider().isClustered() && context.getProperty(HAZELCAST_CLUSTERING_STRATEGY).getValue().equals(CLUSTER_EXPLICIT.getValue())) {
+            results.add(new ValidationResult.Builder()
+                    .subject(HAZELCAST_CLUSTERING_STRATEGY.getName())
+                    .valid(false)
+                    .explanation("Cannot use \"Explicit\" Clustering Strategy when NiFi is not part of a cluster!")
+                    .build());
+        }
+
+        if (!context.getProperty(HAZELCAST_INSTANCES).isSet() && context.getProperty(HAZELCAST_CLUSTERING_STRATEGY).getValue().equals(CLUSTER_EXPLICIT.getValue())) {
+            results.add(new ValidationResult.Builder()
+                    .subject(HAZELCAST_INSTANCES.getName())
+                    .valid(false)
+                    .explanation("In case of \"Explicit\" Clustering Strategy, instances needs to be specified!")
```

Comment:
This error message is not relevant when Explicit strategy cannot be used (non-clustered nifi).
In that case, only the "Cannot use \"Explicit\" Clustering Strategy when NiFi is not part of a cluster" should be displayed.
Or we can keep it, but it should not be the 1st validation error (use List, see my comment in line 183). The 1st should be the "Cannot use \"Explicit\" Clustering Strategy when NiFi is not part of a cluster".

Reply:
I changed the collection into a list so the validation errors will appear in order. As for the conditions I did not change because of this.